augur/filter.py

```diff
@@ -1499,11 +1499,15 @@ def run(args):
                 records_per_group.values(),
                 args.probabilistic_sampling,
             )
-            print(f"Sampling at {sequences_per_group} per group.")
         except TooManyGroupsError as error:
             print(f"ERROR: {error}", file=sys.stderr)
             sys.exit(1)
 
+        if (args.probabilistic_sampling):
+            print(f"Sampling probabilistically at {sequences_per_group:0.4f} sequences per group, meaning it is possible to have more than the requested maximum of {args.subsample_max_sequences} sequences after filtering.")
```

Comment:
Even if the user requests probabilistic sampling, `calculate_sequences_per_group` will try to calculate a deterministic number of sequences per group if it can. This was a past compromise since probabilistic sampling is the default behavior, but when a user requests 10 total sequences, we should try to give them exactly that if we can.

This behavior can lead to odd results in this new message though, where augur ends up not needing to sample probabilistically but the message will still give that impression. You can test this with:

```
augur filter \
  --metadata tests/functional/filter/metadata.tsv \
  --group-by region year month \
  --subsample-max-sequences 10 \
  --output-metadata test.tsv
```

The output from this command is now:

```
Sampling probabilistically at 10.0000 sequences per group, meaning it is possible to have more than the requested maximum of 10 sequences after filtering.
```

When the alternate message would be more accurate:

```
Sampling at 10 sequences per group.
```

Since `calculate_sequences_per_group` is an internal utility, I had originally thought to push print statements up to the calling code in `run`, but we could move these conditional prints down into `calculate_sequences_per_group` (in the corresponding parts of the `try`/`except` block). Alternately, we could change the API of `calculate_sequences_per_group` to return an additional boolean indicating whether it ended up using probabilistic sampling or not.

Do you have a preference one way or the other?

Reply:
Thanks for pointing this out. This makes sense and is probably often the case since `--probabilistic-sampling=True` by default.

I'm with you on scoping print statements to `run`, so returning the additional boolean makes sense. Implemented here: https://github.com/nextstrain/augur/pull/792/commits/6784697a16e16ffc35a995e5aadf873acda8fe06

However, this now has me thinking that the `probabilistic` flag name is a bit misleading. If I'm understanding correctly, it's not whether probabilistic sampling _is_ used, but if it _can be_ used. For example, this variable name change would make more sense: https://github.com/nextstrain/augur/commit/01354d27f3420c085b0a0d650ad07f2c1e6cfc54 but I'm not sure if we want to rename the `--probabilistic-sampling` flag itself as users may already be familiar with it.

Reply:
That's a great point about the misleading nature of the `--probabilistic-sampling` flag. We could clarify the description of that flag in its argparse definition and that would probably be enough. That change would be within the scope of this PR without changing the API of `filter`.

Reply:
Clarified probabilistic sampling: https://github.com/nextstrain/augur/pull/792/commits/26a7c801ee2efb2227f3fafa5b1e3ff548141a23, https://github.com/nextstrain/augur/pull/792/commits/8e44321cb1b37471f67571560ee957f5e3e4e44f

Added tests for output based on original discussion: https://github.com/nextstrain/augur/pull/792/commits/ca1a54b6a83c3ab17f8a53d40a4361ee64854c9f